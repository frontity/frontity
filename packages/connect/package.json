{
  "name": "@frontity/connect",
  "version": "1.1.3",
  "description": "The state manager of Frontity. Similar to MobX but much more lightweight because it's based on ES6 proxies.",
  "keywords": [
    "frontity",
    "connect",
    "state-manager"
  ],
  "homepage": "https://frontity.org",
  "license": "Apache-2.0",
  "main": "src/index.js",
  "types": "src/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/frontity/frontity.git"
  },
  "bugs": {
    "url": "https://community.frontity.org"
  },
  "scripts": {
    "test:ci": "../../node_modules/.bin/jest --ci --coverage",
    "test": "../../node_modules/.bin/jest --watch",
    "test:inspect": "node --inspect ../../node_modules/.bin/jest --watch --no-cache --runInBand"
  },
  "dependencies": {
    "@frontity/error": "^0.1.1"
  },
  "devDependencies": {
    "react-test-renderer": "^16.13.1"
  }
}
