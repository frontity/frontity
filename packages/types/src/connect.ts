/* eslint-disable jsdoc/require-jsdoc */

/**
 * TSDocs disabled by `@luisherranz` in this file because he needs to figure
 * out first which types will remain when we merge this PR
 * https://github.com/frontity/frontity/pull/415, taking into account that
 * `react-easy-state` and `@nx-js/observer-util` already have types themselves.
 */

import { ComponentProps, FunctionComponent, ComponentClass } from "react";
import Package from "./package";
import { ResolveState, ResolveActions, FilterInjectedProps } from "./utils";

export type Connect<Pkg extends Package, Props extends object = {}> = Omit<
  Pkg,
  "state" | "actions" | "name" | "roots" | "fills"
> & {
  state: ResolveState<Pkg["state"]>;
  actions: ResolveActions<Pkg["actions"]>;
} & Props;

export interface ConnectFunction {
  <Comp extends FunctionComponent<any>>(
    comp: Comp,
    options?: { injectProps: boolean }
  ): FunctionComponent<FilterInjectedProps<ComponentProps<Comp>>>;
  <Comp extends ComponentClass<any>>(
    comp: Comp,
    options?: { injectProps: boolean }
  ): ComponentClass<FilterInjectedProps<ComponentProps<Comp>>>;
}

export interface CreateStore {
  <Pkg extends Package>(pkg: Pkg): Omit<Pkg, "state" | "actions"> & {
    state: ResolveState<Pkg["state"]>;
    actions: ResolveActions<Pkg["actions"]>;
  };
}

export default Connect;
