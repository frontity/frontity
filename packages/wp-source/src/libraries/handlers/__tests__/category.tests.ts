import { createStore, InitializedStore, observe } from "@frontity/connect";
import clone from "clone-deep";
import wpSource from "../../../";
import WpSource from "../../../../types";
import Api from "../../api";
// JSON mocks
import { mockResponse } from "./mocks/helpers";
import cat1 from "./mocks/category/cat-1.json";
import cat1Posts from "./mocks/category/cat-1-posts.json";
import cat1PostsPage2 from "./mocks/category/cat-1-posts-page-2.json";
import cat1PostsCpt from "./mocks/category/cat-1-posts-cpt.json";

let store: InitializedStore<WpSource>;
let api: jest.Mocked<Api>;
beforeEach(() => {
  store = createStore(clone(wpSource()));
  store.state.source.api = "https://test.frontity.org/wp-json";
  store.actions.source.init();
  api = store.libraries.source.api as jest.Mocked<Api>;
});

describe("category", () => {
  test("doesn't exist in source.category", async () => {
    // Mock Api responses
    api.get = jest
      .fn()
      .mockResolvedValueOnce(mockResponse([cat1]))
      .mockResolvedValueOnce(
        mockResponse(cat1Posts, {
          "X-WP-Total": "5",
          "X-WP-TotalPages": "2",
        })
      );
    // Fetch entities
    await store.actions.source.fetch("/category/cat-1/");
    expect(store.state.source).toMatchSnapshot();
  });

  test("was populated but not accessed", async () => {
    // Add category to the store
    await store.libraries.source.populate({
      state: store.state,
      response: mockResponse([cat1]),
    });
    // Mock Api responses
    api.get = jest.fn().mockResolvedValueOnce(
      mockResponse(cat1PostsPage2, {
        "X-WP-Total": "5",
        "X-WP-TotalPages": "2",
      })
    );
    // Observe changes in isFetching and isReady properties
    const dataState = [];
    observe(() => {
      const { isFetching, isReady } = store.state.source.get(
        "/category/cat-1/page/2/"
      );
      dataState.push({ isFetching, isReady });
    });
    // Fetch entities
    await store.actions.source.fetch("/category/cat-1/page/2/");
    expect(api.get).toHaveBeenCalledTimes(1);
    expect(store.state.source).toMatchSnapshot();
    // Values history of isFetching and isReady
    expect(dataState).toEqual([
      { isFetching: false, isReady: false }, // first values are from a different object
      { isFetching: true, isReady: false }, // fetch starts
      { isFetching: false, isReady: true }, // fetch ends
    ]);
  });

  test("fetchs from a different endpoint with extra params", async () => {
    // Add custom post endpoint and params
    store.state.source.postEndpoint = "multiple-post-type";
    store.state.source.params = { type: ["post", "cpt"] };
    // Mock Api responses
    api.get = jest
      .fn()
      .mockResolvedValueOnce(mockResponse([cat1]))
      .mockResolvedValueOnce(
        mockResponse(cat1PostsCpt, {
          "X-WP-Total": "5",
          "X-WP-TotalPages": "2",
        })
      );
    // Fetch entities
    await store.actions.source.fetch("/category/cat-1/");
    expect(api.get.mock.calls).toMatchSnapshot();
    expect(store.state.source).toMatchSnapshot();
  });

  test("returns 404 if category doesn't exist in WP", async () => {
    // Mock Api responses
    api.get = jest.fn().mockResolvedValue(mockResponse([]));
    // Fetch entities
    await store.actions.source.fetch("/category/non-existent/");
    expect(api.get).toHaveBeenCalledTimes(1);
    expect(store.state.source).toMatchSnapshot();
  });

  test("returns 404 if the page fetched is out of range", async () => {
    // Mock Api responses
    api.get = jest
      .fn()
      .mockResolvedValueOnce(mockResponse([cat1]))
      .mockResolvedValueOnce(mockResponse([]));
    // Fetch entities
    await store.actions.source.fetch("/category/cat-1/page/3");
    expect(store.state.source).toMatchSnapshot();
  });

  test("doesn't return 404 if the first page is empty", async () => {
    // Mock Api responses
    api.get = jest
      .fn()
      .mockResolvedValueOnce(mockResponse([cat1]))
      .mockResolvedValueOnce(
        mockResponse([], {
          "X-WP-Total": "0",
          "X-WP-TotalPages": "0",
        })
      );
    // Fetch entities
    await store.actions.source.fetch("/category/cat-1/");
    expect(store.state.source).toMatchSnapshot();
  });

  test("doesn't return 404 if the first page is empty (no headers)", async () => {
    // Mock Api responses
    api.get = jest
      .fn()
      .mockResolvedValueOnce(mockResponse([cat1]))
      .mockResolvedValueOnce(mockResponse([], {}));
    // Fetch entities
    await store.actions.source.fetch("/category/cat-1/");
    expect(store.state.source).toMatchSnapshot();
  });

  test("is requested with any query param", async () => {
    // Mock Api responses
    api.get = jest
      .fn()
      .mockResolvedValueOnce(mockResponse([cat1]))
      .mockResolvedValueOnce(
        mockResponse(cat1Posts, {
          "X-WP-Total": "5",
          "X-WP-TotalPages": "2",
        })
      );
    // Fetch entities
    await store.actions.source.fetch("/category/cat-1/?some=param");
    expect(store.state.source).toMatchSnapshot();
  });

  test("Unknown URL should return a 404 even if it's substring matches a path", async () => {
    api.get = jest.fn((_) =>
      Promise.resolve(
        mockResponse(cat1Posts, {
          "X-WP-Total": "5",
          "X-WP-TotalPages": "2",
        })
      )
    );

    await store.actions.source.fetch("/category/undefined/cat-1/");

    expect(
      (store.state.source.data as any)["/category/undefined/cat-1/"]
        .errorStatusText
    ).toBe(
      "You have tried to access content at route: /category/undefined/cat-1/ but it does not exist"
    );

    expect(store.state.source).toMatchSnapshot();
  });
});
