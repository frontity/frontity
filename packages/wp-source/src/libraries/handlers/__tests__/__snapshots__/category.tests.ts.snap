// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`category Unknown URL should return a 404 even if it's substring matches a path 1`] = `
Object {
  "api": "https://test.frontity.org/wp-json",
  "attachment": Object {
    "1": Object {
      "author": 1,
      "id": 1,
      "link": "/post-1/attachment-1/",
      "slug": "attachment-1",
      "type": "attachment",
    },
    "2": Object {
      "author": 1,
      "id": 2,
      "link": "/post-2/attachment-2/",
      "slug": "attachment-2",
      "type": "attachment",
    },
    "3": Object {
      "author": 1,
      "id": 3,
      "link": "/post-3/attachment-3/",
      "slug": "attachment-3",
      "type": "attachment",
    },
  },
  "author": Object {
    "1": Object {
      "id": 1,
      "link": "/author/author-1/",
      "name": "Author 1",
      "slug": "author-1",
    },
  },
  "authorBase": "",
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "data": Object {
    "/author/author-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
    },
    "/category/cat-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "taxonomy": "category",
    },
    "/category/undefined/cat-1/": Object {
      "errorStatus": 404,
      "errorStatusText": "You have tried to access content at route: /category/undefined/cat-1/ but it does not exist",
      "is404": true,
      "isError": true,
      "isFetching": false,
      "isReady": true,
    },
    "/post-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-1/attachment-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
    "/post-2/": Object {
      "id": 2,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-2/attachment-2/": Object {
      "id": 2,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
    "/post-3/": Object {
      "id": 3,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-3/attachment-3/": Object {
      "id": 3,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {},
  "post": Object {
    "1": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          1,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 1,
      "id": 1,
      "link": "/post-1/",
      "slug": "post-1",
      "tags": Array [],
      "type": "post",
    },
    "2": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          2,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 2,
      "id": 2,
      "link": "/post-2/",
      "slug": "post-2",
      "tags": Array [],
      "type": "post",
    },
    "3": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          3,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 3,
      "id": 3,
      "link": "/post-3/",
      "slug": "post-3",
      "tags": Array [],
      "type": "post",
    },
  },
  "postEndpoint": "posts",
  "postTypes": Array [],
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
  "taxonomies": Array [],
  "taxonomy": Object {},
  "type": Object {},
}
`;

exports[`category doesn't exist in source.category 1`] = `
Object {
  "api": "https://test.frontity.org/wp-json",
  "attachment": Object {
    "1": Object {
      "author": 1,
      "id": 1,
      "link": "/post-1/attachment-1/",
      "slug": "attachment-1",
      "type": "attachment",
    },
    "2": Object {
      "author": 1,
      "id": 2,
      "link": "/post-2/attachment-2/",
      "slug": "attachment-2",
      "type": "attachment",
    },
    "3": Object {
      "author": 1,
      "id": 3,
      "link": "/post-3/attachment-3/",
      "slug": "attachment-3",
      "type": "attachment",
    },
  },
  "author": Object {
    "1": Object {
      "id": 1,
      "link": "/author/author-1/",
      "name": "Author 1",
      "slug": "author-1",
    },
  },
  "authorBase": "",
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "data": Object {
    "/author/author-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
    },
    "/category/cat-1/": Object {
      "id": 1,
      "isArchive": true,
      "isCategory": true,
      "isFetching": false,
      "isReady": true,
      "isTaxonomy": true,
      "items": Array [
        Object {
          "id": 1,
          "link": "/post-1/",
          "type": "post",
        },
        Object {
          "id": 2,
          "link": "/post-2/",
          "type": "post",
        },
        Object {
          "id": 3,
          "link": "/post-3/",
          "type": "post",
        },
      ],
      "link": "/category/cat-1/",
      "next": "/category/cat-1/page/2/",
      "page": 1,
      "query": Object {},
      "route": "/category/cat-1/",
      "taxonomy": "category",
      "total": 5,
      "totalPages": 2,
    },
    "/post-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-1/attachment-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
    "/post-2/": Object {
      "id": 2,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-2/attachment-2/": Object {
      "id": 2,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
    "/post-3/": Object {
      "id": 3,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-3/attachment-3/": Object {
      "id": 3,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {},
  "post": Object {
    "1": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          1,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 1,
      "id": 1,
      "link": "/post-1/",
      "slug": "post-1",
      "tags": Array [],
      "type": "post",
    },
    "2": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          2,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 2,
      "id": 2,
      "link": "/post-2/",
      "slug": "post-2",
      "tags": Array [],
      "type": "post",
    },
    "3": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          3,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 3,
      "id": 3,
      "link": "/post-3/",
      "slug": "post-3",
      "tags": Array [],
      "type": "post",
    },
  },
  "postEndpoint": "posts",
  "postTypes": Array [],
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
  "taxonomies": Array [],
  "taxonomy": Object {},
  "type": Object {},
}
`;

exports[`category doesn't return 404 if the first page is empty (no headers) 1`] = `
Object {
  "api": "https://test.frontity.org/wp-json",
  "attachment": Object {},
  "author": Object {},
  "authorBase": "",
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "data": Object {
    "/category/cat-1/": Object {
      "id": 1,
      "isArchive": true,
      "isCategory": true,
      "isFetching": false,
      "isReady": true,
      "isTaxonomy": true,
      "items": Array [],
      "link": "/category/cat-1/",
      "page": 1,
      "query": Object {},
      "route": "/category/cat-1/",
      "taxonomy": "category",
      "total": 0,
      "totalPages": 0,
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {},
  "post": Object {},
  "postEndpoint": "posts",
  "postTypes": Array [],
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
  "taxonomies": Array [],
  "taxonomy": Object {},
  "type": Object {},
}
`;

exports[`category doesn't return 404 if the first page is empty 1`] = `
Object {
  "api": "https://test.frontity.org/wp-json",
  "attachment": Object {},
  "author": Object {},
  "authorBase": "",
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "data": Object {
    "/category/cat-1/": Object {
      "id": 1,
      "isArchive": true,
      "isCategory": true,
      "isFetching": false,
      "isReady": true,
      "isTaxonomy": true,
      "items": Array [],
      "link": "/category/cat-1/",
      "page": 1,
      "query": Object {},
      "route": "/category/cat-1/",
      "taxonomy": "category",
      "total": 0,
      "totalPages": 0,
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {},
  "post": Object {},
  "postEndpoint": "posts",
  "postTypes": Array [],
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
  "taxonomies": Array [],
  "taxonomy": Object {},
  "type": Object {},
}
`;

exports[`category fetchs from a different endpoint with extra params 1`] = `
Array [
  Array [
    Object {
      "endpoint": "categories",
      "params": Object {
        "slug": "cat-1",
      },
    },
  ],
  Array [
    Object {
      "endpoint": "multiple-post-type",
      "params": Object {
        "_embed": true,
        "categories": 1,
        "page": 1,
        "search": undefined,
        "type": Array [
          "post",
          "cpt",
        ],
      },
    },
  ],
]
`;

exports[`category fetchs from a different endpoint with extra params 2`] = `
Object {
  "api": "https://test.frontity.org/wp-json",
  "attachment": Object {
    "1": Object {
      "author": 1,
      "id": 1,
      "link": "/post-1/attachment-1/",
      "slug": "attachment-1",
      "type": "attachment",
    },
    "12": Object {
      "author": 1,
      "id": 12,
      "link": "/cpt/cpt-11/attachment-cpt-11/",
      "slug": "attachment-cpt-11",
      "type": "attachment",
    },
    "2": Object {
      "author": 1,
      "id": 2,
      "link": "/post-2/attachment-2/",
      "slug": "attachment-2",
      "type": "attachment",
    },
  },
  "author": Object {
    "1": Object {
      "id": 1,
      "link": "/author/author-1/",
      "name": "Author 1",
      "slug": "author-1",
    },
  },
  "authorBase": "",
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "cpt": Object {
    "11": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          12,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 12,
      "id": 11,
      "link": "/cpt/cpt-11/",
      "slug": "cpt-11",
      "tags": Array [],
      "type": "cpt",
    },
  },
  "data": Object {
    "/author/author-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
    },
    "/category/cat-1/": Object {
      "id": 1,
      "isArchive": true,
      "isCategory": true,
      "isFetching": false,
      "isReady": true,
      "isTaxonomy": true,
      "items": Array [
        Object {
          "id": 1,
          "link": "/post-1/",
          "type": "post",
        },
        Object {
          "id": 11,
          "link": "/cpt/cpt-11/",
          "type": "cpt",
        },
        Object {
          "id": 2,
          "link": "/post-2/",
          "type": "post",
        },
      ],
      "link": "/category/cat-1/",
      "next": "/category/cat-1/page/2/",
      "page": 1,
      "query": Object {},
      "route": "/category/cat-1/",
      "taxonomy": "category",
      "total": 5,
      "totalPages": 2,
    },
    "/cpt/cpt-11/": Object {
      "id": 11,
      "isFetching": false,
      "isReady": false,
      "type": "cpt",
    },
    "/cpt/cpt-11/attachment-cpt-11/": Object {
      "id": 12,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
    "/post-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-1/attachment-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
    "/post-2/": Object {
      "id": 2,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-2/attachment-2/": Object {
      "id": 2,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {
    "type": Array [
      "post",
      "cpt",
    ],
  },
  "post": Object {
    "1": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          1,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 1,
      "id": 1,
      "link": "/post-1/",
      "slug": "post-1",
      "tags": Array [],
      "type": "post",
    },
    "2": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          2,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 2,
      "id": 2,
      "link": "/post-2/",
      "slug": "post-2",
      "tags": Array [],
      "type": "post",
    },
  },
  "postEndpoint": "multiple-post-type",
  "postTypes": Array [],
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
  "taxonomies": Array [],
  "taxonomy": Object {},
  "type": Object {},
}
`;

exports[`category is requested with any query param 1`] = `
Object {
  "api": "https://test.frontity.org/wp-json",
  "attachment": Object {
    "1": Object {
      "author": 1,
      "id": 1,
      "link": "/post-1/attachment-1/",
      "slug": "attachment-1",
      "type": "attachment",
    },
    "2": Object {
      "author": 1,
      "id": 2,
      "link": "/post-2/attachment-2/",
      "slug": "attachment-2",
      "type": "attachment",
    },
    "3": Object {
      "author": 1,
      "id": 3,
      "link": "/post-3/attachment-3/",
      "slug": "attachment-3",
      "type": "attachment",
    },
  },
  "author": Object {
    "1": Object {
      "id": 1,
      "link": "/author/author-1/",
      "name": "Author 1",
      "slug": "author-1",
    },
  },
  "authorBase": "",
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "data": Object {
    "/author/author-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
    },
    "/category/cat-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "taxonomy": "category",
    },
    "/category/cat-1/?some=param": Object {
      "id": 1,
      "isArchive": true,
      "isCategory": true,
      "isFetching": false,
      "isReady": true,
      "isTaxonomy": true,
      "items": Array [
        Object {
          "id": 1,
          "link": "/post-1/",
          "type": "post",
        },
        Object {
          "id": 2,
          "link": "/post-2/",
          "type": "post",
        },
        Object {
          "id": 3,
          "link": "/post-3/",
          "type": "post",
        },
      ],
      "link": "/category/cat-1/?some=param",
      "next": "/category/cat-1/page/2/?some=param",
      "page": 1,
      "query": Object {
        "some": "param",
      },
      "route": "/category/cat-1/",
      "taxonomy": "category",
      "total": 5,
      "totalPages": 2,
    },
    "/post-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-1/attachment-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
    "/post-2/": Object {
      "id": 2,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-2/attachment-2/": Object {
      "id": 2,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
    "/post-3/": Object {
      "id": 3,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-3/attachment-3/": Object {
      "id": 3,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {},
  "post": Object {
    "1": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          1,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 1,
      "id": 1,
      "link": "/post-1/",
      "slug": "post-1",
      "tags": Array [],
      "type": "post",
    },
    "2": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          2,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 2,
      "id": 2,
      "link": "/post-2/",
      "slug": "post-2",
      "tags": Array [],
      "type": "post",
    },
    "3": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          3,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 3,
      "id": 3,
      "link": "/post-3/",
      "slug": "post-3",
      "tags": Array [],
      "type": "post",
    },
  },
  "postEndpoint": "posts",
  "postTypes": Array [],
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
  "taxonomies": Array [],
  "taxonomy": Object {},
  "type": Object {},
}
`;

exports[`category returns 404 if category doesn't exist in WP 1`] = `
Object {
  "api": "https://test.frontity.org/wp-json",
  "attachment": Object {},
  "author": Object {},
  "authorBase": "",
  "category": Object {},
  "categoryBase": "",
  "data": Object {
    "/category/non-existent/": Object {
      "errorStatus": 404,
      "errorStatusText": "entity from endpoint \\"categories\\" with slug \\"non-existent\\" not found",
      "is404": true,
      "isError": true,
      "isFetching": false,
      "isReady": true,
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {},
  "post": Object {},
  "postEndpoint": "posts",
  "postTypes": Array [],
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
  "taxonomies": Array [],
  "taxonomy": Object {},
  "type": Object {},
}
`;

exports[`category returns 404 if the page fetched is out of range 1`] = `
Object {
  "api": "https://test.frontity.org/wp-json",
  "attachment": Object {},
  "author": Object {},
  "authorBase": "",
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "data": Object {
    "/category/cat-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "taxonomy": "category",
    },
    "/category/cat-1/page/3/": Object {
      "errorStatus": 404,
      "errorStatusText": "category with slug \\"cat-1\\" doesn't have page 3",
      "is404": true,
      "isError": true,
      "isFetching": false,
      "isReady": true,
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {},
  "post": Object {},
  "postEndpoint": "posts",
  "postTypes": Array [],
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
  "taxonomies": Array [],
  "taxonomy": Object {},
  "type": Object {},
}
`;

exports[`category was populated but not accessed 1`] = `
Object {
  "api": "https://test.frontity.org/wp-json",
  "attachment": Object {
    "4": Object {
      "author": 1,
      "id": 4,
      "link": "/post-4/attachment-4/",
      "slug": "attachment-4",
      "type": "attachment",
    },
    "5": Object {
      "author": 1,
      "id": 5,
      "link": "/post-5/attachment-5/",
      "slug": "attachment-5",
      "type": "attachment",
    },
  },
  "author": Object {
    "1": Object {
      "id": 1,
      "link": "/author/author-1/",
      "name": "Author 1",
      "slug": "author-1",
    },
  },
  "authorBase": "",
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "data": Object {
    "/author/author-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
    },
    "/category/cat-1/": Object {
      "id": 1,
      "isFetching": false,
      "isReady": false,
      "taxonomy": "category",
    },
    "/category/cat-1/page/2/": Object {
      "id": 1,
      "isArchive": true,
      "isCategory": true,
      "isFetching": false,
      "isReady": true,
      "isTaxonomy": true,
      "items": Array [
        Object {
          "id": 4,
          "link": "/post-4/",
          "type": "post",
        },
        Object {
          "id": 5,
          "link": "/post-5/",
          "type": "post",
        },
      ],
      "link": "/category/cat-1/page/2/",
      "page": 2,
      "previous": "/category/cat-1/",
      "query": Object {},
      "route": "/category/cat-1/",
      "taxonomy": "category",
      "total": 5,
      "totalPages": 2,
    },
    "/post-4/": Object {
      "id": 4,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-4/attachment-4/": Object {
      "id": 4,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
    "/post-5/": Object {
      "id": 5,
      "isFetching": false,
      "isReady": false,
      "type": "post",
    },
    "/post-5/attachment-5/": Object {
      "id": 5,
      "isFetching": false,
      "isReady": false,
      "type": "attachment",
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {},
  "post": Object {
    "4": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          4,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 4,
      "id": 4,
      "link": "/post-4/",
      "slug": "post-4",
      "tags": Array [],
      "type": "post",
    },
    "5": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          5,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 5,
      "id": 5,
      "link": "/post-5/",
      "slug": "post-5",
      "tags": Array [],
      "type": "post",
    },
  },
  "postEndpoint": "posts",
  "postTypes": Array [],
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
  "taxonomies": Array [],
  "taxonomy": Object {},
  "type": Object {},
}
`;
